File: 2023/day05/python/solution.py
from pathlib import Path


def parse_input(text: str):
    """Parse input into seeds and list of maps."""
    sections = text.strip().split('\n\n')

    # Parse seeds
    seeds = list(map(int, sections[0].split(': ')[1].split()))

    # Parse maps
    maps = []
    for section in sections[1:]:
        lines = section.strip().split('\n')
        ranges = []
        for line in lines[1:]:  # Skip header
            dst_start, src_start, length = map(int, line.split())
            ranges.append((dst_start, src_start, length))
        maps.append(ranges)

    return seeds, maps


def apply_map(value: int, ranges: list) -> int:
    """Apply a single map to transform a value."""
    for dst_start, src_start, length in ranges:
        if src_start <= value < src_start + length:
            return dst_start + (value - src_start)
    return value


def seed_to_location(seed: int, maps: list) -> int:
    """Convert a seed number to a location number through all maps."""
    value = seed
    for map_ranges in maps:
        value = apply_map(value, map_ranges)
    return value


def part1(seeds: list, maps: list) -> int:
    """Find the lowest location number for any initial seed."""
    return min(seed_to_location(seed, maps) for seed in seeds)


def apply_map_to_ranges(input_ranges: list, map_ranges: list) -> list:
    """Apply a map to a list of ranges, returning new ranges."""
    result = []

    for start, end in input_ranges:
        remaining = [(start, end)]

        for dst_start, src_start, length in map_ranges:
            src_end = src_start + length
            new_remaining = []

            for r_start, r_end in remaining:
                # Part before the map range (unmapped)
                if r_start < src_start:
                    new_remaining.append((r_start, min(r_end, src_start)))

                # Part within the map range (mapped)
                overlap_start = max(r_start, src_start)
                overlap_end = min(r_end, src_end)
                if overlap_start < overlap_end:
                    offset = dst_start - src_start
                    result.append((overlap_start + offset, overlap_end + offset))

                # Part after the map range (unmapped)
                if r_end > src_end:
                    new_remaining.append((max(r_start, src_end), r_end))

            remaining = new_remaining

        # Any remaining parts are unmapped (identity)
        result.extend(remaining)

    return result


def part2(seeds: list, maps: list) -> int:
    """Find the lowest location for seed ranges."""
    # Convert seeds to ranges: pairs of (start, start + length)
    ranges = []
    for i in range(0, len(seeds), 2):
        start, length = seeds[i], seeds[i + 1]
        ranges.append((start, start + length))

    # Apply each map to the ranges
    for map_ranges in maps:
        ranges = apply_map_to_ranges(ranges, map_ranges)

    # Find minimum start of any range
    return min(start for start, end in ranges)


def main():
    input_path = Path(__file__).parent.parent / 'input.txt'
    text = input_path.read_text()

    seeds, maps = parse_input(text)

    print('Part 1:', part1(seeds, maps))
    print('Part 2:', part2(seeds, maps))


if __name__ == '__main__':
    main()

File: 2023/day05/rust/solution.rs
use std::fs;

/// A mapping range: (destination_start, source_start, length)
type MapRange = (u64, u64, u64);

/// A complete map is a list of ranges
type Map = Vec<MapRange>;

/// An interval range for Part 2: (start, end) where end is exclusive
type Interval = (u64, u64);

/// Parse the input into seeds and a list of maps
fn parse_input(text: &str) -> (Vec<u64>, Vec<Map>) {
    let sections: Vec<&str> = text.trim().split("\n\n").collect();

    // Parse seeds
    let seeds: Vec<u64> = sections[0]
        .split(": ")
        .nth(1)
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    // Parse maps
    let maps: Vec<Map> = sections[1..]
        .iter()
        .map(|section| {
            section
                .lines()
                .skip(1) // Skip header line
                .map(|line| {
                    let nums: Vec<u64> = line
                        .split_whitespace()
                        .map(|s| s.parse().unwrap())
                        .collect();
                    (nums[0], nums[1], nums[2])
                })
                .collect()
        })
        .collect();

    (seeds, maps)
}

/// Apply a single map to transform a value
fn apply_map(value: u64, ranges: &Map) -> u64 {
    for &(dst_start, src_start, length) in ranges {
        if src_start <= value && value < src_start + length {
            return dst_start + (value - src_start);
        }
    }
    value
}

/// Convert a seed number to a location number through all maps
fn seed_to_location(seed: u64, maps: &[Map]) -> u64 {
    let mut value = seed;
    for map_ranges in maps {
        value = apply_map(value, map_ranges);
    }
    value
}

/// Find the lowest location number for any initial seed (Part 1)
fn part1(seeds: &[u64], maps: &[Map]) -> u64 {
    seeds
        .iter()
        .map(|&seed| seed_to_location(seed, maps))
        .min()
        .unwrap()
}

/// Apply a map to a list of intervals, returning new intervals
fn apply_map_to_ranges(input_ranges: &[Interval], map_ranges: &Map) -> Vec<Interval> {
    let mut result = Vec::new();

    for &(start, end) in input_ranges {
        let mut remaining = vec![(start, end)];

        for &(dst_start, src_start, length) in map_ranges {
            let src_end = src_start + length;
            let mut new_remaining = Vec::new();

            for (r_start, r_end) in remaining {
                // Part before the map range (unmapped)
                if r_start < src_start {
                    new_remaining.push((r_start, r_end.min(src_start)));
                }

                // Part within the map range (mapped)
                let overlap_start = r_start.max(src_start);
                let overlap_end = r_end.min(src_end);
                if overlap_start < overlap_end {
                    let offset = dst_start as i64 - src_start as i64;
                    result.push((
                        (overlap_start as i64 + offset) as u64,
                        (overlap_end as i64 + offset) as u64,
                    ));
                }

                // Part after the map range (unmapped)
                if r_end > src_end {
                    new_remaining.push((r_start.max(src_end), r_end));
                }
            }

            remaining = new_remaining;
        }

        // Any remaining parts are unmapped (identity)
        result.extend(remaining);
    }

    result
}

/// Find the lowest location for seed ranges (Part 2)
fn part2(seeds: &[u64], maps: &[Map]) -> u64 {
    // Convert seeds to ranges: pairs of (start, start + length)
    let mut ranges: Vec<Interval> = seeds
        .chunks(2)
        .map(|chunk| (chunk[0], chunk[0] + chunk[1]))
        .collect();

    // Apply each map to the ranges
    for map_ranges in maps {
        ranges = apply_map_to_ranges(&ranges, map_ranges);
    }

    // Find minimum start of any range
    ranges.iter().map(|&(start, _)| start).min().unwrap()
}

fn main() {
    let input_path = std::path::Path::new(file!())
        .parent()
        .unwrap()
        .join("../input.txt");
    let text = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| fs::read_to_string("../input.txt").unwrap());

    let (seeds, maps) = parse_input(&text);

    println!("Part 1: {}", part1(&seeds, &maps));
    println!("Part 2: {}", part2(&seeds, &maps));
}
